src/components/application_manager/src/request_tracker.cc

```diff
@@ -0,0 +1,127 @@
+
+namespace application_manager {
+
+namespace request_controller {
+
+CREATE_LOGGERPTR_GLOBAL(logger_, "RequestController")
+
+RequestTracker::RequestTracker(const RequestControlerSettings& settings)
+    : settings_(settings) {}
+
+TrackResult RequestTracker::Track(const ApplicationID& app_id,
+                                  const mobile_apis::HMILevel::eType level) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  bool track_result = false;
+
+  LOG4CXX_DEBUG(logger_, "Tracking request for level: " << level);
+
+  if (mobile_apis::HMILevel::HMI_NONE == level) {
+    track_result = Track(app_id,
+                         settings_.app_hmi_level_none_time_scale(),
+                         settings_.app_hmi_level_none_time_scale_max_requests(),
+                         none_level_tracker_);
+
+    return track_result ? kSuccess : kNoneLevelMaxRequestsExceeded;
+  }
+
+  track_result = Track(app_id,
+                       settings_.app_time_scale(),
+                       settings_.app_time_scale_max_requests(),
+                       tracker_);
+
+  return track_result ? kSuccess : kMaxRequestsExceeded;
+}
+
+bool RequestTracker::Track(const ApplicationID& app_id,
+                           const uint32_t time_scale,
+                           const uint32_t max_requests,
+                           ApplicationsRequestsTracker& tracker) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  if (!time_scale || !max_requests) {
+    LOG4CXX_INFO(logger_, "Time scale request tracking is disabled.");
+    return true;
+  }
+
+  LOG4CXX_DEBUG(logger_,
+                "Time scale is: " << time_scale << ". Max requests number is: "
+                                  << max_requests);
+
+  ApplicationsRequestsTracker::iterator it_app = tracker.find(app_id);
+
+  LOG4CXX_DEBUG(logger_, "Tracking app id: " << app_id);
+  if (tracker.end() == it_app) {
+    LOG4CXX_DEBUG(logger_, "Adding new application into tracking.");
+    tracker[app_id].push_back(date_time::DateTime::getCurrentTime());
+    return true;
+  }
+
+  LOG4CXX_DEBUG(logger_,
+                "Amount of known requests is: " << it_app->second.size());
+
+  if (it_app->second.size() < max_requests) {
+    LOG4CXX_DEBUG(logger_, "Adding new request into tracking.");
+    tracker[app_id].push_back(date_time::DateTime::getCurrentTime());
+    return true;
+  }
+
+  LOG4CXX_DEBUG(
+      logger_,
+      "Oldest request is added at: "
+          << date_time::DateTime::getmSecs(it_app->second.front())
+          << ". Current time is: " << date_time::DateTime::getmSecs(
+                                          date_time::DateTime::getCurrentTime())
+          << ". Time scale is: " << time_scale);
+
+  if (date_time::DateTime::calculateTimeSpan(it_app->second.front()) >
+      time_scale) {
+    LOG4CXX_DEBUG(logger_, "Dropping oldest request, adding new one.");
+    ApplicationsRequestsTracker::mapped_type& times = tracker[app_id];
+
+    times.erase(times.begin());
```

Comment:
@dev-gh it shouldn't be empty in ideal world, but this one is not. Some any bugs can be caused after further changes of code, note that tracker is not thread safe, so it is possible to have some even race conditions here. But at least please add DCHEK for explicit pointing that it cant be empty.